### PR TITLE
Add BuildItems that can be used to control instrumentation behavior in dev-mode

### LINKED ISSUE #15330: Allow to skip bytecode instrumentation and force a full reload for specific classes
**Description**

Quarkus 1.11 introduced bytecode instrumentation to avoid reloading of unchanged classes when not strictly necessary. This improvement doesn't work well with Kogito for the following 2 reasons:

**1. Missing reactivity to changes to an external non-java resource**

**When a resource A referenced within a class B is changed, the class B is not reloaded.**

Our DMN implementation has a Java class that, when loaded, parses as part of static initialization an external resource, with extension .dmn, and stores the result of this parsing into a static field. In Kogito extension we have a CompilationProvider implementation that is notified by Quarkus when a .dmn file is changed and when this happens it regenerates the java class with that static field. 

Although a .dmn change notifies the Kogito extension properly, the bytecode of the newly generated class is identical to the old one (e.g.: “load file model1.dmn” code remains the same). However, before the Quarkus change that I linked above, what Quarkus did during hot reload was reloading the new class which caused the external resource, the only file that actually changed, to be indeed parsed again. Since Quarkus 1.11 it realizes that the bytecode of the regenerated class isn’t changed and then it doesn’t perform the class reloading thus ignoring the changes in the dmn file.

We have been able to consistently reproduce the change of Quarkus devmode hot reload behaviour, and we have been able to find a workaround, by introducing a synthetic static field which will be different at each code generation. We wonder however if it’s possible to explicitly mark a specific class to be reloaded regardless.

**2. Class redefinition failed: attempted to change the class NestHost or NestMembers attribute**

For what regards rules, .drl files, the situation is totally different. When a drl file is changed we regenerate Java classes that are different from the previous version but when these classes are loaded into Quarkus something goes wrong. In particular during the first hot reload everything seems to work even though I see in the terminal an exception saying `java.lang.UnsupportedOperationException: class redefinition failed: attempted to change the class NestHost or NestMembers attribute`. 

A second change in the drl file triggers another hot reload that breaks everything triggering an exception saying org.jboss.resteasy.spi.UnhandledException: java.lang.NoClassDefFoundError: org/mfusco/P01/LambdaPredicate0183C8E09EBA30A9B359FFDB99FF4F85

**Implementation ideas**

For these reasons we need a way to tell Quarkus that for classes generated by the Kogito extension we need to skip bytecode instrumentation and force a full reload regardless of the content of the bytecode.

Introduce a new BuildItem allowing to provide the class names that should skip instrumentation and result in a full reload.

### PR DESCRIPTION
**Pull request by geoand**
Fixes: #15330

### CODE REVIEW

**Comment by geoand** [comment]
@mariofusco care to give this a shot?

File: core/deployment/src/main/java/io/quarkus/deployment/dev/ConfigureDisableInstrumentationBuildStep.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        if (processor != null) {
+            processor.setDisableInstrumentationForIndexPredicate(determineEffectivePredicate(forIndexItems))
+                    .setDisableInstrumentationForClassPredicate(determineEffectivePredicate(forClassItems));
+        }
+        return null;
+    }
+
+    private <T> Predicate<T> determineEffectivePredicate(List<? extends Supplier<Predicate<T>>> suppliers) {
+        if (suppliers.isEmpty()) {
+            return new AlwaysFalsePredicate<>();
```
Comment (aloubyansky) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/ConfigureDisableInstrumentationBuildStep.java:29]: Looks like this one could be a singleton.

Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/ConfigureDisableInstrumentationBuildStep.java:29]: Yeah, it initially was, but then the code doesn't look good as the singleton doesn't play well with generics

**Comment by geoand** [comment]
@mariofusco @evacchi were you able to test this?

**Comment by mariofusco** [comment]
@geoand for Kogito extension we found a better solution that, as anticipated in [this email](https://groups.google.com/g/quarkus-dev/c/IA3CdczoRhg), mainly implies removing that early return during hot reload in our build step.

In essence we no longer need this feature right now, but I cannot exclude we will in future and also I find it of general usefulness regardless of Kogito's needs. My suggestion is to merge this pull request and make this feature available to quarkus users anyway.

**Comment by geoand** [comment]
Gotcha, thanks @mariofusco !

File: core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java [hunk 2]
```diff
@@ -222,11 +217,12 @@ public boolean doScan(boolean userInitiated) throws IOException {
                     }
                     Index current = indexer.complete();
                     boolean ok = instrumentationEnabled()
-                            && !containsStartupCode(current);
+                            && !disableInstrumentationForIndexPredicate.test(current);
```
Comment (aloubyansky) [inline core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java:220]: Do we really want to replace the containsStartupCode here? It looks like start up code is an exception anyway.

Comment (aloubyansky) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java:220]: Ah, sorry, you moved it.

**Review by aloubyansky — APPROVED** [review]

File: core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java (hunk 2 above)
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/dev/RuntimeUpdatesProcessor.java:220]: Exactly :)

**Comment by geoand** [comment]
@gsmet made the point the other day that we should probably do the same thing we do for `@Startup` to `@PostConstruct`. 

With the PR, that becomes trivial

**Pull request merged**